mollie-payments-for-woocommerce/includes/mollie/wc/gateway/abstract.php

```diff
@@ -1763,23 +2034,71 @@ protected function get_recurring_total() {
 
 			if ( ! empty( WC()->cart->recurring_carts ) ) {
 
-				$this->recurring_total = array (); // Reset for cached carts
+				$this->recurring_totals = array (); // Reset for cached carts
 
 				foreach ( WC()->cart->recurring_carts as $cart ) {
 
 					if ( ! $cart->prices_include_tax ) {
-						$this->recurring_total[] = $cart->cart_contents_total;
+						$this->recurring_totals[] = $cart->cart_contents_total;
 					} else {
-						$this->recurring_total[] = $cart->cart_contents_total + $cart->tax_total;
+						$this->recurring_totals[] = $cart->cart_contents_total + $cart->tax_total;
 					}
 				}
 			} else {
 				return false;
 			}
 		}
 
-		return $this->recurring_total;
+		return $this->recurring_totals;
+	}
+
+	/**
+	 * Get available payment methods in checkout based on amount, currency and sequenceType
+	 *
+	 * @param $filters
+	 * @return bool
+	 */
+	protected function getAvailableMethodsInCheckout( $filters ) {
+
+		$settings_helper = Mollie_WC_Plugin::getSettingsHelper();
+		$test_mode       = $settings_helper->isTestModeEnabled();
+
+		try {
+			$methods = Mollie_WC_Plugin::getApiHelper()->getApiClient( $test_mode )->methods->all( $filters );
```

Comment:
In the checkout (see `is_available`) we call this API for each payment gateway with the same `$filters`. We call the API even more times when WC Subscriptions is enabled. Is there a way to cache the response and use the response between the different payment gateways?

Reply:
Caching was implemented in v3, but with Guzzle API results can't be serialised, so removed for now and asked Willem, he opened this issue: 
https://github.com/mollie/mollie-api-php/issues/212

Do you think this is a blocker or okay for now?

There is a (hacky) workaround:
https://github.com/mollie/WooCommerce/commit/7e1e06d6bc9517efe0fa6f011b92e6ee8cdec2e4

Reply:
I put this after the methods API call:

```
Mollie_WC_Plugin::debug($this->id . " - Get Mollie methods with filters: " . json_encode($filters) . ' - response ' . json_encode($methods));
```

When I open the checkout I get this in my log:

```
2018-06-25T13:54:39+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:39+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:39+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:39+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:39+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:39+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:40+00:00 DEBUG mollie_wc_gateway_bitcoin - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:40+00:00 DEBUG mollie_wc_gateway_creditcard - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:40+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:40+00:00 DEBUG mollie_wc_gateway_ideal - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:40+00:00 DEBUG mollie_wc_gateway_bancontact - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_bitcoin - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_creditcard - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_directdebit - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_ideal - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
2018-06-25T13:54:42+00:00 DEBUG mollie_wc_gateway_bancontact - Get Mollie methods with filters: {"amount":{"currency":"USD","value":"145.00"},"sequenceType":"oneoff"} - response {}
```

I don't even have all gateways enabled. I understand it's difficult to cache because of the Guzzle changes and the fact you have different input parameters for each checkout (amount). But can the different gateways not somehow share the response? 

I also don't understand why this method is called so many times for `directdebit`.

Reply:
Maybe you can use a `static $variable` to hold the response during the request so it can be shared between the different payment gateways.

Reply:
I'll give this another try and see what I can do about it!

Reply:
Keep in mind, WooCommerce calls `is_available()` at least twice per checkout page load.

> I don't even have all gateways enabled. 

Current version only checks payment methods that are enabled in WooCommerce. hen you say `enabled` do you mean in WooCommerce or Mollie dashboard?

> I understand it's difficult to cache because of the Guzzle changes and the fact you have different input parameters for each checkout (amount).

I've re-implemented cache based on the discussion here:
https://github.com/mollie/mollie-api-php/issues/212

I've extended the `$transient_id` to also include currency, value, and sequenceType. Example:
_transient_mollie-wc-api_methods_test_EUR_75535_first

Per match with the cache (filters to transient ID), the API will now be called once for `oneoff` payments and up to thrice for cart with `oneoff` and `recurring` payments (`oneoff`, `first`, `recurring`). The rest comes from the cache. 

> I also don't understand why this method is called so many times for directdebit.

Checking. I think it might be this, testing now:
https://github.com/mollie/WooCommerce/blob/master/mollie-payments-for-woocommerce/includes/mollie/wc/gateway/abstractseparecurring.php#L16-L20